clients/drcachesim/simulator/cache_misses_analyzer.cpp

```diff
@@ -0,0 +1,159 @@
+    if (misses_count_threshold > kMissesCountThreshold) {
+        misses_count_threshold = kMissesCountThreshold;
+    }
+
+    // Find loads that should be analyzed and analyze them.
+    std::vector<prefetching_recommendation_t*> recommendations;
+    for (auto& pc_cache_misses_it : pc_cache_misses) {
+        std::vector<addr_t>& cache_misses = pc_cache_misses_it.second;
+
+        if (cache_misses.size() >= misses_count_threshold) {
+            const int64_t stride = check_for_constant_stride(cache_misses);
+            if (stride != 0) {
+                prefetching_recommendation_t *recommendation =
+                    new prefetching_recommendation_t;
+                recommendation->pc = pc_cache_misses_it.first;
+                recommendation->stride = stride;
+                recommendation->locality = "nta";
+                recommendations.push_back(recommendation);
+            }
+        }
+    }
+
+    return recommendations;
+}
+
+int64_t
+cache_misses_stats_t::check_for_constant_stride(
+    const std::vector<addr_t>& cache_misses) const
+{
+    std::map<int64_t, int64_t> stride_counts;
+
+    // Find and count all strides in the misses stream.
+    for (int i = 1; i < cache_misses.size(); ++i) {
+        int64_t stride = cache_misses[i] - cache_misses[i - 1];
+        if (stride != 0) {
+          stride_counts[stride]++;
+        }
+    }
+
+    // Find the most occurring stride.
+    int64_t max_count = 0;
+    int64_t max_count_stride = 0;
+    for (auto& stride_count : stride_counts) {
+        if (stride_count.second > max_count) {
+            max_count = stride_count.second;
+            max_count_stride = stride_count.first;
+        }
+    }
+
+    // Return the most occurring stride if it meets the confidence threshold.
+    stride_counts.clear();
+    if (max_count >= static_cast<int64_t>(kConfidenceThreshold * cache_misses.size())) {
+        return max_count_stride;
+    } else {
+        return 0;
+    }
+}
+
+cache_misses_analyzer_t::cache_misses_analyzer_t(const cache_simulator_knobs_t& knobs,
+                                                 uint64_t misses_count_threshold,
+                                                 double misses_frac_threshold,
+                                                 double confidence_threshold)
+    : cache_simulator_t(knobs)
+{
+  if (!success) {
+    return;
+  }
+  bool warmup_enabled = (knobs.warmup_refs > 0 || knobs.warmup_fraction > 0.0);
+
+  delete llcaches["LL"]->get_stats();
+  ll_stats_ = new cache_misses_stats_t(warmup_enabled, knobs.line_size,
+                                       misses_count_threshold,
+                                       misses_frac_threshold,
+                                       confidence_threshold);
+  llcaches["LL"]->set_stats(ll_stats_);
+}
+
+std::vector<prefetching_recommendation_t*>
+cache_misses_analyzer_t::generate_recommendations()
+{
+    return ll_stats_->generate_recommendations();
+}
```

Comment:
I think you forgot to include changes to the cmake code?  These files are never built.

Also, I assume there will be runtime option changes to allow the user to select this: will they be in a separate CL?  That seems fine since the test here will be built and testing the code until then.